app/Kernel/AbstractServer.php

```diff
@@ -0,0 +1,268 @@
+     */
+    protected $app;
+
+    /**
+     * @var string
+     */
+    protected $basePath;
+
+    /**
+     * @var string
+     */
+    protected $publicPath;
+
+    /**
+     * @var string
+     */
+    protected $storagePath;
+
+    /**
+     * @var array
+     */
+    protected $config;
+
+    /**
+     * @var callable[]
+     */
+    protected $extendCallbacks = [];
+
+    /**
+     * @param null $basePath
+     * @param null $publicPath
+     */
+    public function __construct($basePath = null, $publicPath = null)
+    {
+        if ($basePath === null) {
+            $basePath = getcwd();
+        }
+
+        if ($publicPath === null) {
+            $publicPath = $basePath;
+        }
+
+        $this->basePath = $basePath;
+        $this->publicPath = $publicPath;
+
+        if (file_exists($file = $this->basePath.'/config.php')) {
+            $this->config = include $file;
+        }
+    }
+
+    /**
+     * @return string
+     */
+    public function getBasePath()
+    {
+        return $this->basePath;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPublicPath()
+    {
+        return $this->publicPath;
+    }
+
+    /**
+     * @return string
+     */
+    public function getStoragePath()
+    {
+        return $this->storagePath;
+    }
+
+    /**
+     * @param $basePath
+     */
+    public function setBasePath($basePath)
+    {
+        $this->basePath = $basePath;
+    }
+
+    /**
+     * @param $publicPath
+     */
+    public function setPublicPath($publicPath)
+    {
+        $this->publicPath = $publicPath;
+    }
+
+    /**
+     * @param $storagePath
+     */
+    public function setStoragePath($storagePath)
+    {
+        $this->storagePath = $storagePath;
+    }
+
+    /**
+     * @return array
+     */
+    public function getConfig()
+    {
+        return $this->config;
+    }
+
+    /**
+     * @param array $config
+     */
+    public function setConfig(array $config)
+    {
+        $this->config = $config;
+    }
+
+    /**
+     * @param callable $callback
+     */
+    public function extend(callable $callback)
+    {
+        $this->extendCallbacks[] = $callback;
+    }
+
+    /**
+     * @return Application
+     */
+    public function getApp()
+    {
+        if ($this->app !== null) {
+            return $this->app;
+        }
+
+        date_default_timezone_set('UTC');
+
+        $app = new Application($this->basePath, $this->publicPath);
+
+        if ($this->storagePath) {
+            $app->useStoragePath($this->storagePath);
+        }
+
+        $app->instance('env', 'production');
+        $app->instance('mybb.config', $this->config);
+        $app->instance('config', $config = $this->getIlluminateConfig($app));
+
+        $this->registerLogger($app);
+
+        $this->registerCache($app);
+
+        //todo: Register Local, Twig, Database, and other needed providors here
+        $app->register('MyBB\Core\Settings\SettingsServiceProvider');
```

Comment:
We should also probably load these from a config file or something, having them hardcoded here seems off to me.